### PR TITLE
Allow config to be passed into Indexer

### PR DESCRIPTION
**Pull request by andyw8**
(New version of https://github.com/Shopify/ruby-lsp/pull/1436)

### Motivation

We plan to rename `.index.yml` to `.ruby-lsp.yml` and use that for general configuration. As a first step, this PR refactors the indexer to accept a configuration hash, instead of reading the config file directly.

### Implementation

Pass the config as a hash.

### Automated Tests

Updated

### Manual Tests

* Open this branch
* Ensure indexing is working (e.g. the Jump to Definition works for constants)
* Edit `.index.yml` to make it invalid syntax and restart the LSP
* Ensure an error message is showing but otherwise everything works (e.g. formatting)
* Delete `.index.yml` and restart the LSP
* Ensure indexing completes successfully (e.g. the Jump to Definition works for constants)

### CODE REVIEW

**Comment by andyw8** [comment]
(the flaky Windows tests can be ignored, that should be resolved by https://github.com/Shopify/ruby-lsp/pull/1815)

File: lib/ruby_lsp/server.rb [hunk 1]
```diff
@@ -236,7 +236,29 @@ def run_initialized
...
 
-      perform_initial_indexing
+      indexing_config = {}
+
+      # Need to use the workspace URI, otherwise, this will fail for people working on a project that is a symlink.
+      index_path = File.join(@store.workspace_uri.to_standardized_path, ".index.yml")
+
+      if File.exist?(index_path)
+        begin
+          indexing_config = YAML.parse_file(".index.yml").to_ruby
```
Comment (vinistock) [inline lib/ruby_lsp/server.rb:246]:
```suggestion
          indexing_config = YAML.parse_file(index_path).to_ruby
```

File: lib/ruby_lsp/server.rb [hunk 2]
```diff
@@ -236,7 +236,29 @@ def run_initialized
...
+        rescue Psych::SyntaxError => e
+          message = "Syntax error while loading .index.yml configuration: #{e.message}"
```
Comment (vinistock) [inline lib/ruby_lsp/server.rb:248]:
```suggestion
          message = "Syntax error while loading #{index_path} configuration: #{e.message}"
```

**Review by vinistock — APPROVED** [review]

**Review by st0012 — APPROVED** [review]

File: lib/ruby_lsp/server.rb [hunk 3]
```diff
@@ -236,7 +236,29 @@ def run_initialized
...
+      indexing_config = {}
+
+      # Need to use the workspace URI, otherwise, this will fail for people working on a project that is a symlink.
+      index_path = File.join(@store.workspace_uri.to_standardized_path, ".index.yml")
+
+      if File.exist?(index_path)
+        begin
+          indexing_config = YAML.parse_file(index_path).to_ruby
+        rescue Psych::SyntaxError => e
+          message = "Syntax error while loading configuration: #{e.message}"
```
[the remark this answers is not in the document]
Comment (vinistock) [reply lib/ruby_lsp/server.rb:248]: Sounds good

**Pull request merged**